pkg/services/ngalert/notifier/channels/slack.go

```diff
@@ -144,31 +161,85 @@ func (sn *SlackNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool,
 		return false, errors.Wrap(err, "marshal json")
 	}
 
-	cmd := &models.SendWebhookSync{
-		Url:        sn.URL,
-		Body:       string(b),
-		HttpMethod: http.MethodPost,
+	sn.log.Debug("Sending Slack API request", "url", sn.URL.String(), "data", string(b))
+	request, err := http.NewRequestWithContext(ctx, http.MethodPost, sn.URL.String(), bytes.NewReader(b))
+	if err != nil {
+		return false, fmt.Errorf("failed to create HTTP request: %w", err)
 	}
 
-	if sn.Token != "" {
-		sn.log.Debug("Adding authorization header to HTTP request")
-		cmd.HttpHeader = map[string]string{
-			"Authorization": fmt.Sprintf("Bearer %s", sn.Token),
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Set("User-Agent", "Grafana")
+	if sn.Token == "" {
+		if sn.URL.String() == slackAPIEndpoint {
+			panic("Token should be set when using the Slack chat API")
```

Comment:
Can we return an error instead of panic? Also, since we already check this condition during initialising, can we assume this will never happen (by not exporting the struct variables)?

Reply:
@codesome actually using `panic` is the right thing to do here, since if the token is empty at this point it will be a hard programming error and not a user error. The panic is exactly here to avoid assuming that the token was already checked, but instead verify :)

If you wan't I can de-export the struct fields, no problem. But it won't eliminate the possibility of this bug :)

Reply:
Sounds fine since the user can get to this panic without a coding error :)